uc.c

```diff
@@ -1951,85 +2058,174 @@ static void find_context_reg_rw_function(uc_arch arch, uc_mode mode,
+}
+
+UNICORN_EXPORT
+uc_err uc_context_reg_read(uc_context *ctx, int regid, void *value)
+{
+    size_t size = (size_t)-1;
+    return find_context_reg_rw(ctx->arch, ctx->mode)
+        .read(ctx->data, ctx->mode, regid, value, &size);
+}
+
+UNICORN_EXPORT
+uc_err uc_context_reg_write2(uc_context *ctx, int regid, const void *value,
+                             size_t *size)
+{
+    int setpc = 0;
+    return find_context_reg_rw(ctx->arch, ctx->mode)
+        .write(ctx->data, ctx->mode, regid, value, size, &setpc);
+}
+
+UNICORN_EXPORT
+uc_err uc_context_reg_read2(uc_context *ctx, int regid, void *value,
+                            size_t *size)
+{
+    return find_context_reg_rw(ctx->arch, ctx->mode)
+        .read(ctx->data, ctx->mode, regid, value, size);
+}
+
+UNICORN_EXPORT
+uc_err uc_context_reg_write_batch(uc_context *ctx, int *regs, void *const *vals,
                                   int count)
 {
-    int ret = UC_ERR_OK;
-    context_reg_rw_t rw;
+    reg_write_t reg_write = find_context_reg_rw(ctx->arch, ctx->mode).write;
+    void *env = ctx->data;
+    int mode = ctx->mode;
+    int setpc = 0;
+    int i;
 
-    find_context_reg_rw_function(ctx->arch, ctx->mode, &rw);
-    if (rw.context_reg_write) {
-        ret = rw.context_reg_write(ctx, (unsigned int *)ids, vals, count);
-    } else {
-        return UC_ERR_HANDLE;
+    for (i = 0; i < count; i++) {
+        unsigned int regid = regs[i];
+        const void *value = vals[i];
+        size_t size = (size_t)-1;
+        uc_err err = reg_write(env, mode, regid, value, &size, &setpc);
+        if (err) {
+            return err;
+        }
     }
 
-    return ret;
+    return UC_ERR_OK;
 }
 
 UNICORN_EXPORT
-uc_err uc_context_reg_read_batch(uc_context *ctx, int *ids, void **vals,
+uc_err uc_context_reg_read_batch(uc_context *ctx, int *regs, void **vals,
                                  int count)
 {
-    int ret = UC_ERR_OK;
-    context_reg_rw_t rw;
+    reg_read_t reg_read = find_context_reg_rw(ctx->arch, ctx->mode).read;
+    void *env = ctx->data;
+    int mode = ctx->mode;
+    int i;
 
-    find_context_reg_rw_function(ctx->arch, ctx->mode, &rw);
-    if (rw.context_reg_read) {
-        ret = rw.context_reg_read(ctx, (unsigned int *)ids, vals, count);
-    } else {
-        return UC_ERR_HANDLE;
+    for (i = 0; i < count; i++) {
+        unsigned int regid = regs[i];
+        void *value = vals[i];
+        size_t size = (size_t)-1;
+        uc_err err = reg_read(env, mode, regid, value, &size);
+        if (err) {
+            return err;
+        }
     }
 
-    return ret;
+    return UC_ERR_OK;
+}
+
+UNICORN_EXPORT
+uc_err uc_context_reg_write_batch2(uc_context *ctx, int *regs,
+                                   const void *const *vals, size_t *sizes,
+                                   int count)
+{
+    reg_write_t reg_write = find_context_reg_rw(ctx->arch, ctx->mode).write;
+    void *env = ctx->data;
+    int mode = ctx->mode;
+    int setpc = 0;
```

Comment:
Why is `setpc` argument added?

Reply:
I see, for reducing code duplication for `break_translation_loop`, right?

Reply:
Yes, it's to reduce code duplication - specifically, it lets us merge the handling of context and non-context registers, which substantially reduces the amount of per-arch code that needs to be maintained.

At some point, we'll probably need an "isContext" flag so we don't go touching CPU state in context operations, but that's an issue for a future PR.